### ISSUE TITLE
Do Not Resume Paused Partitions After a Rebalance

### ISSUE THREAD

**Issue by garyrussell** [spring-projects/spring-kafka#2222]
If individual partitions are paused (by `@RetryableTopic` or by users), they will be resumed by a rebalance.

While this is not really a problem for `@RetryableTopic`, because the partition will be automatically paused again, but it is a problem for user-paused partitions.

We already handle this case when the whole container is paused, we need something similar for paused partitions.

https://github.com/spring-projects/spring-kafka/blob/02784654a4a32648b807e2b7608af655ff249ea7/spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java#L3272-L3276

**Comment by tomazfernandes** [comment]
You can assign this to me @garyrussell, I should be able to pick it up sometime this week, in time for 2.8.5.

Thanks!

**Comment by garyrussell** [comment]
Thanks; I believe we also need to remove any `pauseRequested` entries for any partitions are are not reassigned to us after a rebalance (in case we get them back on a later rebalance). We need to take into account cooperative rebalancing (where there is only a partial revoke, but we might get some of those back again). I think we need to keep track of which partitions were revoked then, in onPartitionsAssigned, re-pause any newly assigned that are in pauseRequested and remove any revoked and not re-assigned partitions from the pauseRequested.

**Comment by tomazfernandes** [comment]
Ok, sounds a bit more complicated than what I had anticipated 😄 Thanks for the pointers, when I open a PR we can discuss if that looks alright - and please let me know if you think of anything else.

Considering there are some concepts I'm not that familiar with (never heard of cooperative rebalance before, sounds interesting), I'm not sure now if I can deliver it in time for us to properly review and discuss it before the next release. But we've come this far without this, hopefully it won't be a problem if we end up skipping one more version.

Or, sounds like you have a solution figured out already, if you prefer to go ahead and implement it I'm ok with that too.

Thanks!

**Comment by garyrussell** [comment]
OK; I'll take it back; was planning to do it soon anyway.

**Comment by tomazfernandes** [comment]
Sure, thanks! I'll look into these concepts anyway and look forward for your solution - hopefully I'll be able to have a better understanding of it from the pointers you gave.

**Comment by stephprat** [comment]
Hi, I am using 2.8.5 or 2.9.0-RC1 and still having a similar issue in which the KafkaMessageListenerContainers does not have the value of the ConcurrentMessageListenerContainer after a rebalance.
![image](https://user-images.githubusercontent.com/47363068/178707764-c64b2022-194c-420e-a7d3-9d5fa661a4b2.png)

**Comment by garyrussell** [comment]
Oh - right - I can see that could happen if there are two rebalances.

When using Non-blocking retries, the framework will re-pause that partition if the time is not yet satisfied.

When manually pausing individual partitions, that won't happen.

To satisfy both use cases, we must not remove the partitions from the pause requested list when a rebalance occurs.

For non-blocking retries, we MUST remove them, othewise

**Comment by garyrussell** [comment]
Yes, if you add a `ConsumerRebalanceListener`, it will be called by the internal rebalance listener after it has re-paused (and incorrectly removed the revoked partitions from the `pauseRequestedPartitions`), so you can re-request the pauses and, on the next rebalance, when the revoked partition is re-assigned, they will be paused again.

**Comment by stephprat** [comment]
I am using `ConsumerAwareRebalanceListener.onPartitionsAssigned()` which provides directly the consumer and its partitions.
